README.md
# koa-joi-router docs generator

[![NPM version][npm-image]][npm-url]
[![npm download][download-image]][download-url]

[npm-image]: https://img.shields.io/npm/v/@rudi23/koa-joi-router-docs.svg?style=flat-square
[npm-url]: https://www.npmjs.com/package/@rudi23/koa-joi-router-docs
[download-image]: https://img.shields.io/npm/dm/@rudi23/koa-joi-router-docs.svg?style=flat-square
[download-url]: https://www.npmjs.com/package/@rudi23/koa-joi-router-docs
[koa-joi-router]: https://github.com/koa-better-modules/joi-router
[swagger]: https://swagger.io

This project is based on [chuyik/koa-joi-router-docs](https://github.com/chuyik/koa-joi-router-docs).

A node module for generating [Swagger 2.0][swagger] JSON definitions from existing [koa-joi-router][] routes.

## Installation

Install using [`npm`][npm-url]:

```bash
npm install @rudi23/koa-joi-router-docs
```

NodeJS `>= 12.0.0.` is required.

## Example
Visit [example/](./example) folder to see the full example.

## API

#### `new SwaggerAPI()`
Creates a new SwaggerAPI instance.

#### `swaggerAPI.addJoiRouter(router, options)`

Add a joi-router instance to the API. The router should already have all its
routes set up before calling this method (which pulls the route definitions
from the router's `.routes` property).

Options:
- `prefix`: prefix to add to Swagger path (use prefix from JoiRouter if not set)

#### `swaggerAPI.generateSpec(baseSpec, options)`

Create a Swagger specification for this API. A base specification should be
provided with an `info` object (containing at least the `title` and `version`
strings) and any other global descriptions.

Options:
- `defaultResponses`: xustom default responses
  ```js
  {
    200: {
      description: 'Success'
    }
  }
  ```


CHANGELOG.md
## 1.0.0 (2022-07-16)

### Features

- init project ([52d53d7](https://github.com/rudi23/koa-joi-router-docs/commit/52d53d7f3c44a43dd20b7262ee7cdbdad3561e85))

### Bug Fixes

- adjust example ([2c28c24](https://github.com/rudi23/koa-joi-router-docs/commit/2c28c24f66ca3a4fad7d1e3211b35c0f704e2049))
- fix linting ([2c94312](https://github.com/rudi23/koa-joi-router-docs/commit/2c94312244b8e185c0d6c08dd9915ad0c1e5e200))

### Miscellaneous Chores

- add conventional-changelog ([62c0433](https://github.com/rudi23/koa-joi-router-docs/commit/62c0433b536e4948e6770728cf872271bc9fd36f))
- configure release-it ([18dd9fe](https://github.com/rudi23/koa-joi-router-docs/commit/18dd9fefea10a3071dc209f05c4e35166d20364f))
- fix .gitignore ([1700c62](https://github.com/rudi23/koa-joi-router-docs/commit/1700c62f8ec27bfe880d765537a1771d07f90f17))
- improve development with lint-staged and husky, change lint configuration ([282ff8e](https://github.com/rudi23/koa-joi-router-docs/commit/282ff8ececb2b9760cb5a2ccd137c7656569726e))
- pin dependencies ([e64ec18](https://github.com/rudi23/koa-joi-router-docs/commit/e64ec1854c41ca949bdf024701beb655bb909e3e))
- set up codeql ([33b2abf](https://github.com/rudi23/koa-joi-router-docs/commit/33b2abf4a6ebfedd796b36c11f4d630a1bfc1333))
- set up release-it ([0908056](https://github.com/rudi23/koa-joi-router-docs/commit/09080567d62251bd869d6ad904be8019d28a5b2e))
- setup commitlint ([3b82630](https://github.com/rudi23/koa-joi-router-docs/commit/3b82630ecdba884d728085011d3af1e218a0a87a))
- setup github actions ([202fa0d](https://github.com/rudi23/koa-joi-router-docs/commit/202fa0db3ba464c038abaf63b5e5165de65af86c))
- setup github actions ([6bdbc2f](https://github.com/rudi23/koa-joi-router-docs/commit/6bdbc2f57e6cd072e951b4635a9d61bcbd91ce27))
- setup renovate ([3c2819b](https://github.com/rudi23/koa-joi-router-docs/commit/3c2819be6faa72958160cc2a3ef8d3a1b800f7fa))


test/test-api.js
const assert = require('power-assert');
const Router = require('@koa-better-modules/joi-router');
const { SwaggerAPI } = require('../src');

const Joi = Router.Joi;

describe('API', () => {
    it('should success with valid data', () => {
        const generator = new SwaggerAPI();
        const router = new Router();

        router.get('/signup', {
            meta: {
                swagger: {
                    summary: 'User Signup',
                },
            },
            validate: {
                type: 'json',
                body: {
                    username: Joi.string().alphanum().min(3).max(30).required(),
                },
                output: {
                    200: {
                        body: {
                            userId: Joi.string().description('Newly created user id'),
                        },
                    },
                },
            },
            handler: () => {},
        });

        generator.addJoiRouter(router);
        const spec = generator.generateSpec({
            info: {
                title: 'Example API',
                version: '1.1',
            },
            basePath: '/',
        });
        assert(['info', 'basePath', 'swagger', 'paths', 'tags'].every((v) => v in spec));
    });

    it('should success with empty default response', () => {
        const generator = new SwaggerAPI();
        const router = new Router();

        router.get('/empty-default-response', {
            validate: {
                output: {
                    201: {
                        body: {
                            ok: Joi.bool(),
                        },
                    },
                },
            },
            handler: () => {},
        });

        generator.addJoiRouter(router);
        const spec = generator.generateSpec(
            {
                info: {
                    title: 'Example API',
                    version: '1.1',
                },
                basePath: '/',
            },
            { defaultResponses: null }
        );
        assert(!('200' in spec.paths['/empty-default-response'].get.responses));
    });

    it('should success with output placed outside of validate', () => {
        const generator = new SwaggerAPI();
        const router = new Router();

        router.get('/output-outside-validate', {
            validate: {
                output: {
                    201: {
                        body: {
                            ok: Joi.bool(),
                        },
                    },
                },
            },
            handler: () => {},
        });

        generator.addJoiRouter(router);
        const spec = generator.generateSpec({
            info: {
                title: 'Example API',
                version: '1.1',
            },
            basePath: '/',
        });
        assert(['200', '201'].every((v) => v in spec.paths['/output-outside-validate'].get.responses));
    });

    it('should consider the router prefix', () => {
        const generator = new SwaggerAPI();
        const router = new Router();
        router.prefix('/api');

        router.get('/signup', {
            meta: {
                swagger: {
                    summary: 'User Signup',
                },
            },
            validate: {},
            handler: () => {},
        });

        generator.addJoiRouter(router);
        const spec = generator.generateSpec({
            info: {
                title: 'Example API',
                version: '1.1',
            },
            basePath: '/',
        });
        assert(['/api/signup'].every((r) => r in spec.paths));
    });

    it('should consider the prefix option over JoiRouter', () => {
        const generator = new SwaggerAPI();
        const router = new Router();
        router.prefix('/api');

        router.get('/signup', {
            meta: {
                swagger: {
                    summary: 'User Signup',
                },
            },
            validate: {},
            handler: () => {},
        });

        generator.addJoiRouter(router, { prefix: '/other-api' });
        const spec = generator.generateSpec({
            info: {
                title: 'Example API',
                version: '1.1',
            },
            basePath: '/',
        });
        assert(['/other-api/signup'].every((r) => r in spec.paths));
    });

    it('should return $ref with references', () => {
        const generator = new SwaggerAPI();
        const router = new Router();

        const ProfileJoi = Joi.object({
            profileName: Joi.string(),
        });

        router.get('/signup', {
            validate: {
                type: 'json',
                body: ProfileJoi,
                ref: '#/definitions/Profile',
                output: {
                    200: {
                        body: ProfileJoi,
                        ref: '#/definitions/Profile',
                    },
                },
            },
            handler: () => {},
        });

        generator.addJoiRouter(router);
        const spec = generator.generateSpec({
            info: {
                title: 'Example API',
                version: '1.1',
            },
            basePath: '/',
            definitions: {
                Profile: ProfileJoi,
            },
        });
        assert(spec.paths['/signup'].get.parameters[0].schema.$ref === '#/definitions/Profile');
        assert(spec.paths['/signup'].get.responses[200].schema.$ref === '#/definitions/Profile');
    });
});


src/api.js
const assert = require('assert');
const cloneDeep = require('lodash/cloneDeep');
const get = require('lodash/get');
const each = require('lodash/each');
const j2s = require('joi-to-swagger');
const generator = require('./generator');

class SwaggerAPI {
    constructor() {
        this.apiRoutes = [];
    }

    /**
     * Add a `koa-joi-router` router to the API.
     * @param {Router} router - koa-joi-router instance
     * @param {object} options
     * @param {string} [options.prefix]
     */
    addJoiRouter(router, options) {
        options = options || {};

        if (typeof options === 'string') {
            options = { prefix: options };
        }

        if (!Array.isArray(router.routes)) {
            throw new TypeError('router does not have exposed .routes array (not a joi-router instance)');
        }

        const prefix = get(router, 'router.opts.prefix');
        router.routes.forEach(function addRoute(route) {
            this.apiRoutes.push({
                route,
                prefix: options.prefix || prefix,
            });
        }, this);
    }

    /**
     * Generate a Swagger 2.0 specification as an object for this API.
     *
     * @param {object} baseSpec - base document
     * @param {object} baseSpec.info
     * @param {string} baseSpec.info.title
     * @param {string} baseSpec.info.version
     * @param {object} baseSpec.tags
     * @param {string} baseSpec.tags.name
     * @param {string} baseSpec.tags.description
     * @param {object} [options]
     * @param {object} [options.warnFunc]
     * @param {object} [options.defaultResponses]
     * @param {object} [options.definitions]
     * @returns {object} swagger 2.0 specification
     */
    generateSpec(baseSpec, options) {
        options = {
            warnFunc: console.warn,
            defaultResponses: {
                200: {
                    description: 'Success',
                },
                400: {
                    description: 'Bad Request',
                },
                500: {
                    description: 'Failure',
                },
            },
            ...options,
        };

        assert(baseSpec.info, 'baseSpec.info parameter missing');
        assert(baseSpec.info.title, 'baseSpec.info.title parameter missing');
        assert(baseSpec.info.version, 'baseSpec.info.version parameter missing');

        if (!options.defaultResponses) {
            options.defaultResponses = {};
        }

        Object.freeze(options.defaultResponses);

        if (baseSpec.definitions) {
            for (const key in baseSpec.definitions) {
                const val = baseSpec.definitions[key];
                // if it's not a swagger object, convert it
                if (val && val.type !== 'object' && !val.properties) {
                    baseSpec.definitions[key] = j2s(val).swagger;
                }
            }
        }

        const doc = cloneDeep(baseSpec);
        doc.swagger = '2.0';
        doc.paths = doc.paths || {};
        doc.tags = doc.tags || [];

        this.apiRoutes.forEach(function (apiRoute) {
            const routeOptions = { ...options, prefix: apiRoute.prefix };

            /**
             * Caching All `class-validator` Schemas
             */
            try {
                require('reflect-metadata');
                const { getFromContainer, MetadataStorage } = require('class-validator');
                const { validationMetadatasToSchemas } = require('class-validator-jsonschema');
                const metadatas = getFromContainer(MetadataStorage).validationMetadatas;
                const schemas = validationMetadatasToSchemas(metadatas);
                if (schemas) {
                    for (const key in schemas) {
                        const schema = schemas[key];
                        if (schema.required) {
                            each(schema.required, (fieldKey) => {
                                if (schema.properties && schema.properties[fieldKey]) {
                                    schema.properties[fieldKey].required = true;
                                }
                            });
                        }
                    }
                    routeOptions.schemas = schemas;
                }
            } catch (err) {
                if (!/Cannot find module 'class-validator'/.test(err.message)) {
                    throw err;
                }
            }

            const routePaths = generator.routeToSwaggerPaths(apiRoute.route, routeOptions);

            generator.mergeSwaggerPaths(doc.paths, routePaths, options);
        }, this);

        return doc;
    }
}

module.exports = SwaggerAPI;


.release-it.js
module.exports = {
    hooks: {
        'before:bump': ['npm test'],
    },
    git: {
        requireCleanWorkingDir: true,
        requireBranch: 'master',
        commitMessage: 'chore: release ${version} \n\n${changelog}',
        tagName: 'v${version}',
    },
    github: {
        release: true,
    },
    npm: {
        publish: false,
    },
    plugins: {
        '@release-it/conventional-changelog': {
            preset: {
                name: 'conventionalcommits',
                types: [
                    { type: 'feat', section: 'Features' },
                    { type: 'feature', section: 'Features' },
                    { type: 'fix', section: 'Bug Fixes' },
                    { type: 'perf', section: 'Performance Improvements' },
                    { type: 'revert', section: 'Reverts' },
                    { type: 'docs', section: 'Documentation' },
                    { type: 'chore', section: 'Miscellaneous Chores' },
                    { type: 'refactor', section: 'Code Refactoring' },
                    { type: 'test', section: 'Tests' },
                    { type: 'build', section: 'Build System' },
                    { type: 'ci', section: 'Continuous Integration' },
                    { type: 'config', section: 'Configuration changes' },
                ],
            },
            infile: 'CHANGELOG.md',
        },
    },
};


src/generator.js
const pathToRegex = require('path-to-regexp');
const j2s = require('joi-to-swagger');
const each = require('lodash/each');
const pick = require('lodash/pick');
const get = require('lodash/get');
const some = require('lodash/some');
const mapValues = require('lodash/mapValues');

const captureRegex = /(:\w+)/g;

const JSON_SCHEMA_FIELDS = [
    'type',
    'required',

    'maximum',
    'exclusiveMaximum',
    'minimum',
    'exclusiveMinimum',
    'maxLength',
    'minLength',
    /* 'pattern', */ 'maxItems',
    'minItems',
    'uniqueItems',
    'enum',
    'multipleOf',
];

exports.mergeSwaggerPaths = function mergeSwaggerPaths(paths, newPaths, options) {
    const warn = options && options.warnFunc;

    each(newPaths, function (newPathItemObj, path) {
        const pathItemObj = (paths[path] = paths[path] || {});

        // Merge operations into path
        each(newPathItemObj, function (operationObj, method) {
            if (pathItemObj[method]) {
                // already exists!
                if (warn) {
                    warn(`${path}[${method}] exists in multiple routes`);
                }

                return;
            }

            pathItemObj[method] = operationObj;
        });
    });

    return paths;
};

exports.routesToSwaggerPaths = function routesToSwaggerPaths(routes, options) {
    const paths = {};

    routes.forEach(function (route) {
        const routePaths = exports.routeToSwaggerPaths(route, options);

        exports.mergeSwaggerPaths(paths, routePaths, options);
    });

    return paths;
};

/**
 * For a given joi-router route, return an array of Swagger paths.
 * @param {object} route
 * @returns {object} paths
 */
exports.routeToSwaggerPaths = function routeToSwaggerPaths(route, options) {
    options = options || {};

    const paths = {};
    const routeDesc = {
        responses: {},
    };

    if (!route.validate && route.handler && options.schemas) {
        classValidatorToSwagger(route, options.schemas);
    }

    if (route.validate) {
        const type = route.validate.type;

        if (!type) {
            // do nothing
        } else if (type === 'json') {
            routeDesc.consumes = ['application/json'];
        } else if (type === 'form') {
            routeDesc.consumes = ['application/x-www-form-urlencoded'];
        } else if (type === 'multipart') {
            routeDesc.consumes = ['multipart/form-data'];
        }

        routeDesc.parameters = exports.validateToSwaggerParameters(route.validate);

        // add responses from output schema
        const output = route.output || route.validate.output;
        if (output) {
            routeDesc.responses = { ...options.defaultResponses };
            Object.keys(output).forEach((x) => {
                x = x.toString();
                // 200,206,300
                const respCodes = x.split(',');
                respCodes.forEach((code) => {
                    // 200-299
                    if (code.includes('-')) {
                        code = code.split('-')[0];
                    }
                    code = code.trim();
                    routeDesc.responses[code] = {
                        description: 'Success',
                        ...options.defaultResponses[code],
                        ...outputToSwagger(output[x]),
                    };
                });
            });
        }
    }

    if (route.meta && route.meta.swagger) {
        Object.assign(routeDesc, route.meta.swagger);
    }

    // This sets default 'path' parameters so swagger-ui doesn't complain.
    const noPathParamsExist = !some(routeDesc.parameters, ['in', 'path']);
    const noPathValidatorExists = get(route, 'validate.path') === undefined;
    const noParamsValidatorExists = get(route, 'validate.params') === undefined;
    if (noPathParamsExist && noPathValidatorExists && noParamsValidatorExists) {
        routeDesc.parameters = routeDesc.parameters || [];
        const pathCaptures = route.path.match(captureRegex);
        if (pathCaptures) {
            each(pathCaptures, function (pathParameter) {
                routeDesc.parameters.push({
                    name: pathParameter.replace(':', ''),
                    in: 'path',
                    type: 'string',
                    required: true,
                });
            });
        }
    }

    let path = exports.swaggerizePath(route.path);

    if (options.prefix) {
        if (options.prefix.endsWith('/') || path.startsWith('/')) {
            path = `${options.prefix}${path}`;
        } else {
            path = `${options.prefix}/${path}`;
        }
    }

    const pathItemObj = (paths[path] = {});

    const methods = Array.isArray(route.method) ? route.method : [route.method];

    methods.forEach(function (method) {
        const operationObj = routeDesc;
        pathItemObj[method.toLowerCase()] = operationObj;
    });

    return paths;
};

function addSchemaParameters(parameters, location, schema) {
    const swaggerObject = schema.swagger || j2s(schema).swagger;
    if (swaggerObject.type === 'object' && swaggerObject.properties) {
        each(swaggerObject.properties, function (value, name) {
            if (value.type === 'string' && value.format === 'binary') {
                value.type = 'file';
                delete value.format;
            }
            const parameter = value;
            parameter.name = name;
            parameter.in = location;
            parameters.push(parameter);
        });
    }
}

function outputToSwagger(respJson) {
    const obj = {};
    if (respJson && respJson.body) {
        const respBody = respJson.body;
        obj.schema =
            typeof respJson.ref === 'string' ? { $ref: respJson.ref } : respBody.swagger || j2s(respBody).swagger;
        if (respJson.schema || obj.schema.description) {
            obj.description = respJson.schema || obj.schema.description;
        }
    }

    return obj;
}

/**
 * Convert a JSON schema object to the subset used by Swagger
 */
exports.jsonSchemaToSwagger = function jsonSchemaToSwagger(jsonSchema) {
    if (Array.isArray(jsonSchema)) {
        return jsonSchema.map(exports.jsonSchemaToSwagger);
    }

    const schema = pick(jsonSchema, JSON_SCHEMA_FIELDS);

    if (jsonSchema.items) {
        // FIXME HACK
        if (Array.isArray(jsonSchema.items)) {
            jsonSchema.items = jsonSchema.items[0];
        }

        schema.items = exports.jsonSchemaToSwagger(jsonSchema.items);
    }

    if (jsonSchema.properties) {
        schema.properties = mapValues(jsonSchema.properties, function (value) {
            return exports.jsonSchemaToSwagger(value);
        });
    }

    return schema;
};

/**
 * Convert joi-router validate object to swagger
 */
exports.validateToSwaggerParameters = function validateToSwaggerParameters(validate) {
    const parameters = [];

    if (validate.header) {
        addSchemaParameters(parameters, 'header', validate.header);
    }

    if (validate.query) {
        addSchemaParameters(parameters, 'query', validate.query);
    }

    if (validate.path) {
        // Still there for anyone who uses old syntax
        addSchemaParameters(parameters, 'path', validate.path);
    }

    if (validate.params) {
        addSchemaParameters(parameters, 'path', validate.params);
    }

    if (validate.formData) {
        addSchemaParameters(parameters, 'formData', validate.formData);
    }

    if (validate.body) {
        const schema = validate.body;
        const swaggerSchema =
            typeof validate.ref === 'string' ? { $ref: validate.ref } : schema.swagger || j2s(schema).swagger;
        parameters.push({
            name: 'body',
            in: 'body',
            schema: swaggerSchema,
        });
    }

    return parameters;
};

/* Convert a joi-router path into a Swagger parameterized path,
 * e.g. /users/:userId becomes /users/{userId}
 *
 * FIXME: incomplete handling, escaping, etc
 * FIXME: throw error if a complex regex is used
 */
exports.swaggerizePath = function swaggerizePath(path) {
    const pathTokens = pathToRegex.parse(path);

    const segments = pathTokens.map(function (token) {
        let segment = token;

        if (token.name) {
            segment = `{${token.name}}`; // this means this is a complex regex group. for more info, read up on path-to-regexp, koa-joi-router uses it, it's great.
        } else {
            segment = token.replace('/', ''); // remove leading slash, to handle things like complex routes: /users/:userId/friends/:friendId
        }

        return segment;
    });

    return `/${segments.join('/')}`; // path is normalized, just add that leading slash back to handle prefixes properly again.
};

function classValidatorToSwagger(route, schemas) {
    const handler = get(route, 'handler.0');
    if (!handler) {
        return;
    }
    if (!schemas || schemas.length) {
        return;
    }
    try {
        // route --> Controller & method
        const clazz = handler.__class__;
        const method = handler.__method__;
        if (!clazz || !method) {
            return;
        }

        // get configs from @decorators
        const docs = Reflect.getMetadata('Docs', clazz, method);
        const input = Reflect.getMetadata('ValidateInput', clazz, method);
        const output = Reflect.getMetadata('ValidateOutput', clazz, method);

        route.validate = {};
        if (docs) {
            route.meta = { swagger: docs };
        } else {
            route.meta = { swagger: { summary: route.path } };
        }

        if (input) {
            route.validate.type = input.type;

            // class --> class-validator-jsonschema --> swagger schema
            each(['query', 'params', 'body'], (key) => {
                const clazz = input[key];
                if (clazz) {
                    const schema = schemas[clazz.name];
                    if (schema) {
                        route.validate[key] = { swagger: schema };
                    }
                }
            });
        }

        if (output) {
            route.validate.output = {};
            for (const key in output) {
                const schema = schemas[output[key].name];
                if (schema) {
                    route.validate.output[key] = {
                        body: {
                            swagger: schema,
                        },
                    };
                }
            }
        }
    } catch (err) {
        // eslint-disable-next-line no-console
        console.warn(`[joi-router-docs] Fail to generate docs for \`${route.path}\`, reason: `, err);
    }
}


example/routes.js
const Router = require('@koa-better-modules/joi-router');
const { SwaggerAPI } = require('../src');

const Joi = Router.Joi;

/**
 * Define routes
 */
const router = new Router();

// Get /user/:id
router.get('/user/:id', {
    meta: {
        swagger: {
            summary: 'Get User Info',
            description: 'Note: \nSensitive data can only be viewed by the `corresponding user` or `Admin`.',
            tags: ['users'],
        },
    },
    validate: {
        params: {
            id: Joi.string().alphanum().max(24).example('abcdefg').description('User id').required(),
        },
        output: {
            '200-299': {
                body: Joi.object({
                    userId: Joi.string().description('User id'),
                })
                    .options({
                        allowUnknown: true,
                    })
                    .description('User object'),
            },
            500: {
                body: Joi.object({
                    message: Joi.string().description('error message'),
                }).description('error body'),
            },
        },
    },
    handler: async (ctx) => {
        ctx.body = {
            userId: ctx.params.id,
        };
    },
});

// POST /signup
router.post('/signup', {
    meta: {
        swagger: {
            summary: 'User Signup',
            description: 'Signup with username and password.',
            tags: ['users'],
        },
    },
    validate: {
        type: 'json',
        body: Joi.object({
            username: Joi.string().alphanum().min(3).max(30).required(),
            password: Joi.string().alphanum().min(6).max(30).required(),
        }).example({ username: 'abcdefg', password: '123123' }),
        output: {
            200: {
                body: {
                    userId: Joi.string().description('Newly created user id'),
                },
            },
            500: {
                body: {
                    code: Joi.number().description('error code'),
                },
            },
        },
    },
    handler: async (ctx) => {
        ctx.body = {
            userId: ctx.body.username,
        };
    },
});

const ProfileJoi = Joi.object({
    profileName: Joi.string(),
});
// POST /profile using schema references `ref`
router.post('/profile', {
    meta: {
        swagger: {
            summary: 'Profile',
            description: 'Signup with username and password.',
            tags: ['users'],
        },
    },
    validate: {
        type: 'json',
        body: ProfileJoi,
        ref: '#/definitions/Profile',
        output: {
            200: {
                body: ProfileJoi,
                ref: '#/definitions/Profile',
            },
        },
    },
    handler: async (ctx) => {
        ctx.body = {
            profileName: 'this is a profile name',
        };
    },
});

/**
 * Generate Swagger json from the router object
 */
const generator = new SwaggerAPI();
generator.addJoiRouter(router);

const spec = generator.generateSpec(
    {
        info: {
            title: 'Example API',
            description: 'API for creating and editing examples.',
            version: '1.1',
        },
        basePath: '/',
        tags: [
            {
                name: 'users',
                description: `A User represents a person who can login 
      and take actions subject to their granted permissions.`,
            },
        ],
        // pass `definitions` if you need schema references
        definitions: {
            Profile: ProfileJoi,
        },
    },
    {
        defaultResponses: {
            200: {
                description: 'OK',
            },
            500: {
                description: 'ERROR',
            },
        }, // Custom default responses if you don't like default 200
    }
);

/**
 * Swagger JSON API
 */
router.get('/_api.json', async (ctx) => {
    ctx.body = JSON.stringify(spec, null, '  ');
});

/**
 * API documentation
 */
router.get('/apiDocs', async (ctx) => {
    ctx.body = `
  <!DOCTYPE html>
  <html>
  <head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta http-equiv="X-UA-Compatible" content="ie=edge">
    <title>Example API</title>
  </head>
  <body>
    <redoc spec-url='/_api.json' lazy-rendering></redoc>
    <script src="https://rebilly.github.io/ReDoc/releases/latest/redoc.min.js"></script>
  </body>
  </html>
  `;
});

module.exports = router;
